### ISSUE TITLE
[Bug]: Restore precise return types weakened to Any in certification.py and remove type: ignore suppressions

### ISSUE THREAD

**Issue by Adit-Jain-srm** [Adit-Jain-srm/NightmareNet#446]
## Problem

PR #421 (mypy strict mode) introduced two type safety regressions while making modules pass strict checks:

### 1. `certification.py` L158: Return type weakened to `Any`

`_run_noisy_forward_passes` return type changed from `np.ndarray` to `Any` to satisfy strict mode. This loses type safety for all callers - they can no longer rely on the return being an ndarray without runtime checks.

```python
# Before (typed):
def _run_noisy_forward_passes(...) -> np.ndarray:

# After (untyped):
def _run_noisy_forward_passes(...) -> Any:
```

The correct fix is `npt.NDArray[np.int64]` or `np.ndarray[Any, np.dtype[np.int64]]`.

### 2. `prepare.py` L117-118: `type: ignore` suppressions added

Two `# type: ignore[no-untyped-call]` comments were added instead of typing the upstream generators:

```python
dream_data = dream_gen.generate(dream_base)  # type: ignore[no-untyped-call]
nightmare_data = nightmare_gen.generate(nightmare_base)  # type: ignore[no-untyped-call]
```

This hides the real problem (untyped `data/generator.py`) behind suppressions rather than fixing it.

## Acceptance Criteria

- [ ] `certification.py` `_run_noisy_forward_passes` return type restored to `np.ndarray` (or `npt.NDArray[np.int64]`)
- [ ] `_make_noise_hook` return type changed from `Any` to proper hook callable type (e.g., `Callable[[Any, Any, torch.Tensor], torch.Tensor]`)
- [ ] Remove `# type: ignore[no-untyped-call]` from `prepare.py` L117-118 by adding proper type annotations to `data/generator.py`'s `generate()` method
- [ ] All modules still pass mypy strict
- [ ] No test regressions

## Scope

- `nightmarenet/evaluation/certification.py` - restore precise return types
- `nightmarenet/phases/prepare.py` - remove ignore comments (depends on data/ being typed)
- `nightmarenet/data/generator.py` - add `generate()` return type annotation

## Technical Notes

For numpy typing, use:
```python
import numpy.typing as npt

def _run_noisy_forward_passes(...) -> npt.NDArray[np.int64]:
```

For the hook callable:
```python
from typing import Callable, Tuple
HookFn = Callable[[Any, Tuple[Any, ...], torch.Tensor], torch.Tensor]

def _make_noise_hook(sigma: float) -> HookFn:
```

> **Important:** Please read CONTRIBUTING.md in full before requesting assignment. Assignment requests without a technical approach will not be considered. PRs demonstrating thorough research, benchmarks, or creative approaches beyond minimum criteria may earn bonus XP labels.

**Comment by Aryanbuha890** [comment]
Hi @Adit-Jain-srm 

I would like to be assigned to this issue. Here is my proposed technical approach to restore precise types and clean up the suppressions:

### Technical Approach

1. **Restore Return Types in `certification.py`**
   - **`_run_noisy_forward_passes`**: I will import `numpy.typing as npt` and specify the return type as `npt.NDArray[np.int64]` to restore type safety:
     ```python
     import numpy.typing as npt
     import numpy as np

     def _run_noisy_forward_passes(...) -> npt.NDArray[np.int64]:
     ```
   - **`_make_noise_hook`**: I will define a clear signature alias for the PyTorch forward pre/post hook function and annotate the builder return type:
     ```python
     from typing import Callable, Tuple, Any
     import torch

     HookFn = Callable[[Any, Tuple[Any, ...], torch.Tensor], torch.Tensor]

     def _make_noise_hook(sigma: float) -> HookFn:
     ```

2. **Add Types to `data/generator.py` and Cleanup `prepare.py`**
   - I will inspect `generate()` in `data/generator.py` and add proper return type annotations (e.g., matching the data container returned by the dream/nightmare generators).
   - Once `data/generator.py` is fully typed, I will remove the `# type: ignore[no-untyped-call]` annotations from lines 117-118 in `phases/prepare.py` so mypy can rely on clean static type inference.

3. **Validation & Testing**
   - Run local strict checks to ensure no regressions:
     ```bash
     mypy nightmarenet/
     ```
   - Run the test suite to verify everything passes:
     ```bash
     pytest tests/
     ```

Kindly assign this issue to me. I'm ready to get started!

**Comment by Adit-Jain-srm** [comment]
Assigned @Aryanbuha890
